Add test coverage command and update README

pbearne opened the pull request:
Introduced a new `npm run test:test-coverage` command for generating HTML test coverage reports. Updated README to reflect the new command and usage, and modified `.gitignore` to exclude coverage report files.


Trac ticket: https://core.trac.wordpress.org/ticket/53414

Commit:
Add test coverage command and update README

Introduced a new `npm run test:test-coverage` command for generating HTML test coverage reports. Updated README to reflect the new command and usage, and modified `.gitignore` to exclude coverage report files.

desrosj commented on tools/local-env/scripts/docker.js:
```diff
@@ -7,5 +7,10 @@ dotenvExpand.expand( dotenv.config() );
 const composeFiles = local_env_utils.get_compose_files();
 
+if (process.argv.includes('--coverage-html')) {
+	process.env.LOCAL_PHP_XDEBUG = 'true';
```
I haven't tested to confirm, but I don't think this will work as is. Just changing the environment variables has no effect unless the container is stopped and restarted because these are read on environment start up.

pbearne replied:
I wasn't sure, but it seems to work

desrosj replied:
Ah, yes. Confirming that this does work. The reason is that `test:php` technically starts a separate PHP container than any that is already started. So whether the original PHP container has xDebug enabled does not matter.

desrosj commented on package.json:
```diff
@@ -190,6 +190,7 @@
 		"env:pull": "node ./tools/local-env/scripts/docker.js pull",
 		"test:performance": "wp-scripts test-playwright --config tests/performance/playwright.config.js",
 		"test:php": "node ./tools/local-env/scripts/docker.js run php ./vendor/bin/phpunit",
+		"test:test-coverage": "npm run test:php -- --coverage-html ./coverage",
```
```suggestion
		"test:coverage": "npm run test:php -- --coverage-html ./coverage/html/ --coverage-php ./coverage/php/report.php --coverage-text ./coverage/text/report.txt",
```

The reports should go in the `coverage` folder with a directory for each type.

desrosj commented on README.md:
````diff
@@ -94,6 +94,23 @@ npm run test:php -- --filter <test name>
+These command run to create HTML coverage report:
+
+```
+npm run test:test-coverage
````
```suggestion
npm run test:coverage
```

desrosj commented on README.md:
````diff
@@ -94,6 +94,23 @@ npm run test:php -- --filter <test name>
 npm run test:php -- --group <group name or ticket number>
 ```
 
+#### To run the coverage report
````
```suggestion
#### Generating a code coverage report
```

desrosj commented on README.md:
````diff
@@ -94,6 +94,23 @@ npm run test:php -- --filter <test name>
 ```
 
+#### To run the coverage report
+An online version can be found at https://app.codecov.io/gh/WordPress/wordpress-develop
````
```suggestion
Code coverage reports are [generated daily](https://github.com/WordPress/wordpress-develop/actions/workflows/test-coverage.yml) and [submitted to Codecov.io](https://app.codecov.io/gh/WordPress/wordpress-develop).
```

desrosj reviewed (changes requested):
Thanks for refreshing this! Here is some initial feedback.

pbearne committed:
Update package.json

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

pbearne committed:
Update README.md

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

pbearne committed:
Update README.md

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

pbearne committed:
Update README.md

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

Commit:
Fix typo in test coverage command examples

Corrected the npm test coverage command in the README.md to avoid confusion. This change ensures that users run the correct command for generating test coverage.

desrosj commented on tools/local-env/scripts/docker.js:
```diff
@@ -7,5 +7,10 @@ dotenvExpand.expand( dotenv.config() );
 
+if (process.argv.includes('--coverage-html')) {
+	process.env.LOCAL_PHP_XDEBUG = 'true';
+	process.env.LOCAL_PHP_XDEBUG_MODE = process.env.LOCAL_PHP_XDEBUG_MODE + ',coverage';
```
```suggestion
	process.env.LOCAL_PHP_XDEBUG_MODE = 'coverage';
```

No need to preserve any other values for LOCAL_PHP_XDEBUG_MODE`.

desrosj commented on .gitignore:
```diff
@@ -105,3 +105,6 @@ wp-tests-config.php
 
 # Visual regression test diffs
 tests/visual-regression/specs/__snapshots__
+
```
This can be just added in the `# Files and folders related to build/test tools` section near the top without a new comment.

pbearne replied:
moved

desrosj reviewed (commented):
Added a few more things. I am goign to take a pass at reworking the README.md file changes directly.

desrosj commented on package.json:
```diff
@@ -190,6 +190,7 @@
 		"env:pull": "node ./tools/local-env/scripts/docker.js pull",
 		"test:performance": "wp-scripts test-playwright --config tests/performance/playwright.config.js",
 		"test:php": "node ./tools/local-env/scripts/docker.js run php ./vendor/bin/phpunit",
+		"test:coverage": "npm run test:php -- --coverage-html ./coverage/html/ --coverage-php ./coverage/php/report.php --coverage-text ./coverage/text/report.txt",
```
```suggestion
		"test:coverage": "npm run test:php -- --coverage-html ./coverage/html/ --coverage-php ./coverage/php/report.php --coverage-text=./coverage/text/report.txt",
```

Seems like the text option needs to be passed like this.

pbearne committed:
Update tools/local-env/scripts/docker.js

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

pbearne committed:
Update package.json

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

pbearne committed:
Update .gitignore

desrosj committed:
Refine README.md file changes.

pbearne committed:
Update .gitignore

desrosj reviewed (approved):
Thanks @pbearne! This is looking good.

desrosj commented:
Merged in https://core.trac.wordpress.org/changeset/59356.
